src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,631 @@
+	DenseMatrix O(_matrix.rows(), rank);
+	sample_gaussian(O);
+			
+	// Compute Sample Matrix of _matrix^T
+	DenseMatrix Y = _matrix.transpose() * std::move(O);
+			
+	// Orthonormalize Y
+	gram_schmidt(Y);
+			
+	// Range(B) = Range(_matrix^T)
+	DenseMatrix B = _matrix * Y;
+			
+	// Gaussian Random Matrix
+	DenseMatrix P(B.cols(), rank);
+	sample_gaussian(P);
+			
+	// Compute Sample Matrix of B
+	DenseMatrix Z = B * P;
+			
+	// Orthonormalize Z
+	gram_schmidt(Z);
+			
+    // Range(C) = Range(B)
+	DenseMatrix C = Z.transpose() * std::move(B); 
+			
+	Eigen::JacobiSVD<DenseMatrix> svdOfC(C, Eigen::ComputeThinV);
+			
+	// C = USV^T
+	// A = Z * U * S * V^T * Y^T()
+	_singularvalues = svdOfC.singularValues();
+	_singularvectors = std::move(Y) * svdOfC.matrixV();
+
+    return SVDComponentsAnnotationData<TransformedT>(std::move(_singularvalues), std::move(_singularvectors));
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  SVDTransformer
+// |
+// ----------------------------------------------------------------------
+template <typename InputT, typename TransformedT>
+SVDTransformer<InputT, TransformedT>::SVDTransformer(TransformedT singularvalues, TransformedT singularvectors) :
+    //no validation here because same variable has been validated in annotation class
+    _singularvalues(std::move(singularvalues)),
+    _singularvectors(std::move(singularvectors)) {
+}
+
+//TODO:
+// template <typename InputT, typename TransformedT>
+// SVDTransformer<InputT, TransformedT>::SVDTransformer(Archive &ar) :
+//     SVDTransformer(
+//         [&ar](void) {
+//             // Version
+//             std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+//             std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+//             if(majorVersion != 1 || minorVersion != 0)
+//                 throw std::runtime_error("Unsupported archive version");
+
+//             // Data:TODO
+//             TransformedT                   singularvalues(Traits<TransformedT>::deserialize(ar));
+//             TransformedT                   singularvectors(Traits<TransformedT>::deserialize(ar));
+
+//             return SVDTransformer<InputT, TransformedT>(std::move(singularvalues), std::move(singularvectors));
+//         }()
+//     ) {
+// }
+
+template <typename InputT, typename TransformedT>
+bool SVDTransformer<InputT, TransformedT>::operator==(SVDTransformer const &other) const {
+    if ((this->_singularvalues - other._singularvectors).norm() > 0.000001f || (this->_singularvalues - other._singularvectors).norm() > 0.000001f)
+        return false;
+
+    return true;
+}
+
+template <typename InputT, typename TransformedT>
+void SVDTransformer<InputT, TransformedT>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data TODO
```

Comment:
>/ Data TODO [](start = 5, length = 11)

Will transformer function w/o training data?

Reply:
transformer function will contain the components obtained from training data(singularvalues and singularvectors in this featurizer). But I don't think ORT need TODO part(serialization/deserialization) @davidbrownellWork is that correct?

Reply:
All Transformers in ORT are instantiated using deserialization.

---
In reply to: [366112663](https://github.com/microsoft/FeaturizersLibrary/pull/64#discussion_r366112663) [](ancestors = 366112663)

Reply:
Will add serialization/deserialization in Traits in another PR

Reply:
We should introduce partially created Featurizers into master; if you implemented matrix serialization in a separate PR, make sure that one goes in before this one.